apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs

```diff
@@ -518,45 +538,48 @@ private void StartAcquisitionTasksIfNecessary()
+                    createSessionBatchMaxSize = Options.CreateSessionMaximumBatchSize;
                 }
 
-                // Determine how many more requests to start.
-                // We want to make sure that if all existing and new requests succeed:
+                // Determine how many more sessions to create.
+                // We want to make sure that if all existing and new requests for session creation succeed:
                 // - All queuing callers will have a session
-                // - The pool will have at least as many entries the options specify
+                // - The pool will have at least as many entries as the options specify
                 // However, we don't want to end up with more than the maximum number of active sessions
                 // in the options.
 
                 // In reality, the current pool size should be 0 if pendingAcquisitionCount is non-zero, and
                 // vice versa. However, we don't strictly enforce this, and the maths works out more simply if
                 // we don't worry about it.
 
-                // How many more requests do we need to make in order to get the pool to the minimum size, after satisfying
-                // pending callers, assuming that all the in-flight requests succeed, and there are no more requests?
-                int newRequestsToSatisfyPool = (pendingAcquisitionCount + minPoolSize) - (poolSize + inFlightRequests);
-                // How many more requests *can* we make without going over the maximum number of active sessions?
-                int maxAvailableRequests = maxActiveSessions - liveOrRequestedSessionCount;
-
-                int actualNewRequests = Math.Min(newRequestsToSatisfyPool, maxAvailableRequests);
+                // How many more sessions do we need to create in order to get the pool to the minimum size, after satisfying
+                // pending callers, assuming that all the in-flight session requests succeed, and there are no more requests?
+                int newSessionsToSatisfyPool = (pendingAcquisitionCount + minPoolSize) - (poolSize + inFlightRequests);
+                // How many more sessions *can* we create without going over the maximum number of active sessions?
+                int maxPossibleNewSessions = maxActiveSessions - liveOrRequestedSessionCount;
 
-                for (int i = 0; i < actualNewRequests; i++)
+                int remainingNewSessions = Math.Min(newSessionsToSatisfyPool, maxPossibleNewSessions);
+                while (remainingNewSessions > 0)
                 {
-                    Parent.ConsumeBackgroundTask(PrepareNewSessionAsync(), "session creation");
+                    int batchSize = Math.Min(remainingNewSessions, createSessionBatchMaxSize);
+                    Parent.ConsumeBackgroundTask(PrepareNewSessionsAsync(batchSize), "session creation");
+                    remainingNewSessions -= batchSize;
```

Comment:
It's possible that BatchCreateSessions will return fewer than the session count asked for. So I would do this computation in the callback in the async function below.

Reply:
I think we can actually do better. We can explicitly call StartSessionCreationTasksIfNecessary whenever a BatchCreateSession returns less sessions than requested. @jskeet, what do you think?

Reply:
I think it's worth doing whatever is simplest here, to be honest. Do we know how often BatchCreateSessions returns fewer sessions than requested? If it's relatively rare, I'd be happy with "this can occasionally be inefficient, but will never result in session starvation".